README.rst
=============================
Django Saas Email Manager
=============================

.. image:: https://badge.fury.io/py/django-saas-email.svg
    :target: https://badge.fury.io/py/django-saas-email

.. image:: https://travis-ci.org/unicorn-supplies/django-saas-email.svg?branch=master
    :target: https://travis-ci.org/unicorn-supplies/django-saas-email

.. image:: https://codecov.io/gh/unicorn-supplies/django-saas-email/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/unicorn-supplies/django-saas-email

An email manager for sending emails with templates, mail history and admin.

Features
--------

* Use of `Mailhog <https://github.com/mailhog/MailHog/>`_ for local email testing.
* Send emails via `django-anymail <https://github.com/anymail/django-anymail>`_ (using `Sendgrid <https://sendgrid.com>`__ as default).
* Serving dynamic HTML E-Mail Templates, editable with `Tinymce <https://github.com/aljosa/django-tinymce>`_.
* Use of the awesome `Transactional email templates <https://github.com/mailgun/transactional-email-templates>`_ from `Mailgun <https://www.mailgun.com>`_.


Documentation
-------------

The full documentation is at https://django-saas-email.readthedocs.io.

Quickstart
----------

**Installation**

Installation with pip::

    pip install django-saas-email


Add it to your `INSTALLED_APPS`:

.. code-block:: python

    INSTALLED_APPS = (
        ...
        'django_saas_email.apps.DjangoSaasEmailConfig',
        ...
    )


**JSONField**

We are using the Postgres JSONField as default. If you installed `psycopg2`, everything should work fine.

If you are using a different database, you also need to install `jsonfield`::

    pip install jsonfield


**Adding the Sendgrid API Key to your settings**

In settings.py::

    SENDGRID_API_KEY=<....>

Other optional settings::

    DJANGO_SAAS_TEST_EMAIL_ADDRESS=youremailfortesting@example.org
    DJANGO_SAAS_FOOTER="""Follow <a href="#" style="font-family: 'Helvetica Neue',Helvetica,Arial,sans-serif; box-sizing: border-box; font-size: 12px; color: #999; text-decoration: underline; margin: 0;">@yourcompany</a> on Twitter"""

**Sending emails**

Basic example::

    from django_saas_email.utils import create_and_send_emails

    context={
        'first_name': 'John',
        'last_name': 'Doe',
    }

    create_and_send_mail(
        template_name="hello_world",
        context=context,
        to_address=john.doe@example.org
    )

This will create an email and send it with Sengrid.

You should use http://premailer.dialect.ca or django-premailer to create Inline CSS in HTML


Running Tests
-------------

Does the code actually work?::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install tox
    (myenv) $ tox



HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2017-07-24)
++++++++++++++++++

* First release on PyPI.


django_saas_email/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11.3 on 2017-07-24 16:57
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion
import django.utils.timezone

try:
    from django.contrib.postgres.fields import JSONField
except:
    from jsonfield import JSONField

import tinymce.models
import uuid


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Mail',
            fields=[
                ('id', models.UUIDField(default=uuid.uuid4, editable=False, primary_key=True, serialize=False, unique=True, verbose_name='ID')),
                ('from_address', models.EmailField(help_text="The 'from' field of the email", max_length=254, verbose_name='Sender email address')),
                ('to_address', models.EmailField(help_text="The 'to' field of the email", max_length=254, verbose_name='Recipient email address')),
                ('delivery_mail_id', models.IntegerField(blank=True, editable=False, help_text='The ID is saved after correct sending', null=True, verbose_name='Unique mail sender ID')),
                ('delivery_status', models.IntegerField(blank=True, editable=False, help_text='The Mail sender status', null=True, verbose_name='Status of Mail sender')),
                ('subject', models.CharField(blank=True, help_text='Subject line for a mail', max_length=500, null=True, verbose_name='Email Subject line')),
                ('context', JSONField(help_text='JSON dump of context dictionary used to fill in templates', verbose_name='Data of email context')),
                ('time_created', models.DateTimeField(default=django.utils.timezone.now, editable=False, help_text='When was the mail created?', verbose_name='Creation time')),
                ('time_sent', models.DateTimeField(blank=True, editable=False, help_text='When was the mail send via the backend?', null=True, verbose_name='Sent time')),
                ('time_delivered', models.DateTimeField(blank=True, editable=False, help_text='Actual delivery time by the email backend', null=True, verbose_name='Delivery time')),
                ('used_backend', models.CharField(blank=True, editable=False, help_text='Which email backend was used for sending?', max_length=128, null=True, verbose_name='E-Mail Backend')),
            ],
        ),
        migrations.CreateModel(
            name='MailManager',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
            ],
        ),
        migrations.CreateModel(
            name='MailTemplate',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(help_text='Template name; a short all-lowercase string', max_length=100, unique=True, verbose_name='Template name')),
                ('subject', models.CharField(help_text='A format string like "Hello {}"; required', max_length=200, verbose_name='Email subject line template')),
                ('html_template', tinymce.models.HTMLField(help_text="The HTML template, written with Django's template syntax; required", verbose_name='HTML template (required)')),
                ('text_template', models.TextField(blank=True, default='', help_text='This is an optional field for adding a custom text-only version of this template. If left blank, the plaintext email will be generated dynamically from the HTML when needed.', null=True, verbose_name='Text template (optional)')),
            ],
        ),
        migrations.AddField(
            model_name='mail',
            name='template',
            field=models.ForeignKey(help_text='The used template', on_delete=django.db.models.deletion.CASCADE, to='django_saas_email.MailTemplate', verbose_name='Used Mail template'),
        ),
    ]
